pkg/media/opuswriter/opuswriter.go

```diff
@@ -0,0 +1,228 @@
+	"fmt"
+	"io"
+	"math"
+	"math/rand"
+	"os"
+	"time"
+
+	"github.com/pions/rtp"
+	"github.com/pions/rtp/codecs"
+)
+
+// OpusWriter is used to take RTP packets and write them to an OGG on disk
+type OpusWriter struct {
+	stream                  io.Writer
+	fd                      *os.File
+	sampleRate              uint32
+	channelCount            uint16
+	serial                  uint32
+	pageIndex               uint32
+	checksumTable           []int
+	previousGranulePosition uint64
+	previousTimestamp       uint32
+}
+
+// New builds a new OGG Opus writer
+func New(fileName string, sampleRate uint32, channelCount uint16) (*OpusWriter, error) {
+	f, err := os.Create(fileName)
+	if err != nil {
+		return nil, err
+	}
+	writer, err := NewWith(f, sampleRate, channelCount)
+	if err != nil {
+		return nil, err
+	}
+	writer.fd = f
+	return writer, nil
+}
+
+// NewWith initialize a new OGG Opus writer with an io.Writer output
+func NewWith(out io.Writer, sampleRate uint32, channelCount uint16) (*OpusWriter, error) {
+	if out == nil {
+		return nil, fmt.Errorf("file not opened")
+	}
+
+	r := rand.NewSource(time.Now().UnixNano())
+	r1 := rand.New(r)
+	serial := uint32(math.Ceil(r1.Float64() * math.Pow(2, 32)))
+
+	writer := &OpusWriter{
+		stream:        out,
+		sampleRate:    sampleRate,
+		channelCount:  channelCount,
+		serial:        serial,
+		checksumTable: make([]int, 256),
+	}
+	writer.initChecksum()
+	if err := writer.writeHeaders(); err != nil {
+		return nil, err
+	}
+
+	return writer, nil
+}
+
+/*
+    ref: https://tools.ietf.org/html/rfc7845.html
+    https://git.xiph.org/?p=opus-tools.git;a=blob;f=src/opus_header.c#l219
+
+       Page 0         Pages 1 ... n        Pages (n+1) ...
+    +------------+ +---+ +---+ ... +---+ +-----------+ +---------+ +--
+    |            | |   | |   |     |   | |           | |         | |
+    |+----------+| |+-----------------+| |+-------------------+ +-----
+    |||ID Header|| ||  Comment Header || ||Audio Data Packet 1| | ...
+    |+----------+| |+-----------------+| |+-------------------+ +-----
+    |            | |   | |   |     |   | |           | |         | |
+    +------------+ +---+ +---+ ... +---+ +-----------+ +---------+ +--
+    ^      ^                           ^
+    |      |                           |
+    |      |                           Mandatory Page Break
+    |      |
+    |      ID header is contained on a single page
+    |
+    'Beginning Of Stream'
+
+   Figure 1: Example Packet Organization for a Logical Ogg Opus Stream
+*/
+
+func (i *OpusWriter) writeHeaders() error {
+	// ID Header
+	oggIDHeader := make([]byte, 19)
+
+	copy(oggIDHeader[0:], []byte("OpusHead"))                     // Magic Signature 'OpusHead'
+	oggIDHeader[8] = 1                                            // Version
+	oggIDHeader[9] = uint8(i.channelCount)                        // Channel count
+	binary.LittleEndian.PutUint16(oggIDHeader[10:], 0)            // pre-skip, don't need to skip any value
```

Comment:
This is optional because memory defaults to 0, but I like how it's being explicit (even at a tiiiiinny performance hit).